sdk/src/main/scala/io/horizen/account/sc2sc/AbstractCrossChainRedeemMessageProcessor.scala

```diff
@@ -0,0 +1,137 @@
+      AbstractCrossChainMessageProcessor.buildCrosschainMessageFromAccount(accCcRedeemMessage.accountCrossChainMessage, networkParams)
+    )
+    setCrossChainMessageHash(messageHash, view)
+  }
+
+  /**
+   * Extract the AccountCrossChainRedeemMessage from the message
+   *
+   * @param msg message to apply to the state
+   * @return the AccountCrossChainRedeemMessage built from the arguments in the message
+   */
+  protected def getAccountCrossChainRedeemMessageFromMessage(msg: Message): AccountCrossChainRedeemMessage
+
+  private def validateRedeemMsg(ccRedeemMgs: AccountCrossChainRedeemMessage, view: BaseAccountStateView): Unit = {
+    val accountCcMsg = ccRedeemMgs.accountCrossChainMessage
+    // Validate the receiving sidechain matches with this scId
+    validateScId(networkParams.sidechainId, accountCcMsg.receiverSidechain)
+
+    // Validate message has not been redeemed yet
+    val ccMsg = AbstractCrossChainMessageProcessor.buildCrosschainMessageFromAccount(accountCcMsg, networkParams)
+    validateDoubleMessageRedeem(ccMsg, view)
+
+    // Validate scCommitmentTreeRoot and nextScCommitmentTreeRoot exists
+    validateScCommitmentTreeRoot(ccRedeemMgs.scCommitmentTreeRoot, ccRedeemMgs.nextScCommitmentTreeRoot, view)
+
+    // Validate Proof
+    validateProof(ccRedeemMgs)
+  }
+
+  private def validateScId(scId: Array[Byte], receivingScId: Array[Byte]): Unit = {
+    if (!scId.sameElements(receivingScId)) {
+      throw new IllegalArgumentException(s"This scId `${BytesUtils.toHexString(scId)}` and receiving scId `${BytesUtils.toHexString(receivingScId)}` do not match")
+    }
+  }
+
+  private def validateDoubleMessageRedeem(ccMsg: CrossChainMessage, view: BaseAccountStateView): Unit = {
+    val currentMsgHash = sc2scCircuit.getCrossChainMessageHash(ccMsg)
+    val ccMsgFromRedeemAlreadyExists = doesCrossChainMessageHashFromRedeemMessageExist(currentMsgHash, view)
+    if (ccMsgFromRedeemAlreadyExists) {
+      throw new IllegalArgumentException(s"Message $ccMsg has already been redeemed")
+    }
+  }
+
+  override def addScTxCommitmentTreeRootHash(hash: Array[Byte], view: BaseAccountStateView): Unit =
+    view.updateAccountStorage(contractAddress, getScTxCommitmentTreeRootHashKey(hash), Array.emptyByteArray)
+
+  private def validateScCommitmentTreeRoot(
+                                            scCommitmentTreeRoot: Array[Byte],
+                                            nextScCommitmentTreeRoot: Array[Byte],
+                                            view: BaseAccountStateView): Unit = {
+    if (view.getAccountStorage(contractAddress, getScTxCommitmentTreeRootHashKey(scCommitmentTreeRoot)).isEmpty) {
+      throw new IllegalArgumentException(s"Sidechain commitment tree root `${BytesUtils.toHexString(scCommitmentTreeRoot)}` does not exist")
+    }
+
+    if (view.getAccountStorage(contractAddress, getScTxCommitmentTreeRootHashKey(nextScCommitmentTreeRoot)).isEmpty) {
```

Comment:
For checking if a key is defined in the stateDb, this is the check that needs to be done:
`getAccountStorage(.....).sameElements(NULL_HEX_STRING_32)`